### DISCUSSION TITLE
How are you using handle()?

### DISCUSSION THREAD

**Discussion by j4w8n** [sveltejs/kit#6229] [General]
I currently use this function in hooks.js to read browser cookies and set `event.locals`.

I'd love to learn how you're using `handle()` to do rad shit.

👇

**Comment by nikfp** [comment]
Traffic analytics. It's the perfect place. I don't have it fully implemented yet, but I'm using it to ping a webhook with some anonymous data that I can use to watch usage patterns in the app. Since the rest of the request chain isn't dependent on the promise resolving I can send it and move on. Which pages are getting the most use? Which are getting the least? What time of day? Rate of new feature adoption? I'm thinking toward using that data to steer decisions, maybe allow A/B testing, and maybe using a hash of the userID to identify (anonymously) how the top users are using the application. I can also see possibilities for rate limiting and IP blacklisting if working with a cache.

This is all a work in progress though. I'm just playing with it at this point. This is also for an internal application that won't see huge traffic volumes, at least not any time soon.

**Reply by j4w8n** [reply]
Sounds nice!

**Comment by j4w8n** [comment]
![scared-brad-pitt.gif](https://user-images.githubusercontent.com/25783093/186948526-f1df814f-5b84-41f8-a79a-3774eb4f190f.gif)

**Comment by Coretteket** [comment]
I'll jump in onto this thread, mostly because I want to hear from y'all if I'm doing something wrong! 😅

For my work-in-progress [personal site](https://github.com/Coretteket/qntn.io/tree/dev), I use [`hooks.ts`](https://github.com/Coretteket/qntn.io/blob/dev/src/hooks.ts) to read cookie headers for both a user's theme and language preference. If those are unset, I read `accept-language` headers for locales and do some spooky experimental `sec-ch-prefers-color-scheme` stuff for theming. I then modify the response with `transformPageChunk` to set `lang` and `data-theme` properties on the `<html>` tag. Lastly, I pass the theme and locale through `event.locals`.

To use these locals on the client, I currently have a one-line [`layout.server.ts`](https://github.com/Coretteket/qntn.io/blob/dev/src/routes/%2Blayout.svelte) that just passes the `locals` to `data`. Very boilerplate-y, but I haven't figured out a way to do it another way: a lot of the logic I currently have in hooks could also be here, except for changing the `<html>` tag on the server, which means all of it has to stay in hooks. Finally, in [`layout.ts`](https://github.com/Coretteket/qntn.io/blob/dev/src/routes/%2Blayout.ts) I set the theme and locale data to two stores and load the appropriate dictionaries using my own typesafe, routed [`i18n`](https://github.com/Coretteket/qntn.io/tree/dev/src/i18n) implementation.

I would really love to either move my hooks implementation to `layout.server.ts`, or to somehow get rid of the latter entirely. I'd hoped that the routing changes would facilitate this, but it seems my boilerplate has simply moved from `getSession` to `layout.server.ts`. Does anyone have a more elegant solution to this?

**Reply by j4w8n** [reply]
Excellent!

**Reply by binajmen** [reply]
Kind of looking in the same direction, but not sure why you must use `layour.server.ts` to access _hooked_ data in end routes. Could you expand on that?

**Reply by j4w8n** [reply]
@binajmen `locals` is just the way sveltekit has determined to pass data to server-side functions. Good for passing sensitive data like tokens, or non-sensitive data; per the [docs](https://kit.svelte.dev/docs/load).

Coding this "locals pass-through" is a bit boilerplate, and I wish it was automatic, but I'm not sure how many apps need this. One of my demo apps would be fine with it. But I'd think larger apps would still need to pass more data through, other than `locals`.

Even if sveltekit could automatically merge or pass `locals` to `data`, without additional code or a `+layout.server.js` file; this could get dangerous because someone might pass sensitive data to the client side. Why? Because, as you may already know, returning `data` from `+layout.server.js` automatically populates `data` in a `+page.svelte` file which declares `export let data` or imports the `$page` store.

```js
/* hooks.js */

export const handle = async ({ event, resolve }) => {
  event.locals.sensitive = JSON.parse(event.cookies.get('tokens') ?? null)
  event.locals.non_sensitive = JSON.parse(event.cookies.get('user') ?? null)

  const response = await resolve(event)
  return response
}
```

```js
/* pretend this is magically happening behind the scenes, without a +layout.server.js file */

export const load = ({ locals }) => {
  return {
    /* we're now dangerously passing access tokens to the client-side */
    ...locals
  }
}
```

**Comment by accuser** [comment]
I've got several hooks in sequence - request / performance logging, get access / id tokens from cookies, get user claims (from id token), etc. `sequence()` is great for building a pipeline.

**Reply by j4w8n** [reply]
I haven't used sequence yet, but I've seen it around!

**Reply by binajmen** [reply]
Do you have public content you could share? As a SK newcomer, I'm struggling without proper tutorials/examples. It seems hooks will respond to my needs, but I'm not sure what is the impact mentioned by @Coretteket. Do you have also the same boilerplate-y issue?
